addons/mail/static/src/core/common/thread.js

```diff
@@ -396,7 +397,14 @@ export class Thread extends Component {
     }
 
     get PRESENT_THRESHOLD() {
-        return this.state.showJumpPresent ? PRESENT_THRESHOLD - 200 : PRESENT_THRESHOLD;
+        const viewportHeight = this.scrollableRef.el?.clientHeight * MINIMUM_VIEWPORTS_SCROLLED;
+        const messagesHeight = [...this.props.thread.nonEmptyMessages]
+            .reverse()
+            .slice(0, MINIMUM_MESSAGES_SCROLLED)
+            .map((message) => this.refByMessageId.get(message.id))
+            .reduce((totalHeight, message) => totalHeight + message?.el?.clientHeight, 0);
```

Comment:
IMO this part is not very good in regard to time complexity (1 loop for destructuring, 1 for reverse, 1 partial for slice, 1 partial for map, 1 partial for reduce). Moreover, I expect this function to be called quite often as it is a getter that is used in the template.

One approach that would not require any loop would be to:

- Get the message we need directly: `nonEmptyMessages.at(-MINIMUM_MESSAGES_SCROLLED)`.
- Get the corresponding element using `refBymessageId`.
- Compute the height by manipulating the position of the element in its container.

I think something like this should do the trick? However, I haven't thoroughly tested it.


```js
function getTotalHeightToBottom(element) {
    const container = element.parentElement;
    const elRect = element.getBoundingClientRect();
    const bottomRelativeTocontainer = elRect.top - container.getBoundingClientRect().top;
    const totalHeightToBottom =
        container.scrollHeight - container.scrollTop - bottomRelativeTocontainer;
    return totalHeightToBottom;
}
```

Reply:
Honestly I wouldn't bother with making more performant when there's no perceivable performance issues.

Also `getBoundingClientRect()` can actually be a huge performance issue, because this triggers a painting and this is a blocking operation until the painting is finished.